core/src/main/scala/kafka/coordinator/transaction/TransactionCoordinator.scala

```diff
@@ -129,10 +130,45 @@ class TransactionCoordinator(brokerId: Int,
           responseCallback(initTransactionMetadata(metadata))
 
         case Some(metadata) =>
-          metadata synchronized {
-            metadata.epoch = (metadata.epoch + 1).toShort
-          }
-          responseCallback(initTransactionMetadata(metadata))
+          initPidWithExistingMetadata(transactionalId, transactionTimeoutMs, responseCallback, metadata)
+      }
+    }
+  }
+
+  private def initPidWithExistingMetadata(transactionalId: String,
+                                          transactionTimeoutMs: Int,
+                                          responseCallback: InitPidCallback,
+                                          metadata: TransactionMetadata) = {
+
+    metadata synchronized {
+      if (metadata.state == Ongoing) {
+        // abort the ongoing transaction
+        handleEndTransaction(transactionalId,
+          metadata.pid,
+          metadata.epoch,
+          TransactionResult.ABORT,
+          (errors: Errors) => {
+            if (errors != Errors.NONE) {
+              responseCallback(initTransactionError(errors))
+            } else {
+              // init pid again
+              handleInitPid(transactionalId, transactionTimeoutMs, responseCallback)
+            }
+          })
+      } else if (metadata.state == PrepareAbort || metadata.state == PrepareCommit) {
+        // wait for the commit to complete and then init pid again
+        txnMarkerPurgatory.tryCompleteElseWatch(new DelayedTxnMarker(metadata, () => {
+          handleInitPid(transactionalId, transactionTimeoutMs, responseCallback)
+        }), Seq(metadata.pid))
+      } else {
+        // reset the metadata and respond with new epoch
+        // TODO: should this append to the transaction log?, i.e., it doesn't above in the None case
```

Comment:
Not sure I follow this comment. If you are bumping the epoch, the new epoch must go to the transaction log. Otherwise the epoch will 'roll backward` if the coordinator moves. Or does this comment mean something else?